contracts/StorageManager.sol

```diff
@@ -172,177 +156,180 @@ contract StorageManager {
-        numberOfPeriodsDeposited.mul(billingPrice); // overFlow check. If this doesn't pass, we might have issues transfering the maximum amount REF_MAX_TRANSFER
+        require(agreement.lastPayoutDate == 0, "StorageManager: Agreement already active");
+
+        uint64 billingPrice = offer.billingPlans[billingPeriod];
+        require(billingPrice != 0, "StorageManager: billing price doesn't exist for Offer");
+
+        agreement.availableFunds = msg.value;
+        require(agreement.availableFunds >= size * billingPrice, "StorageManager: funds deposited has to be for at least one billing period");
+
+        agreement.size = size;
         agreement.billingPrice = billingPrice;
         agreement.billingPeriod = billingPeriod;
-        agreement.numberOfPeriodsDeposited = uint64(numberOfPeriodsDeposited);
-        agreement.startDate = uint128(now);
+        agreement.lastPayoutDate = uint128(now);
+
+        offer.availableCapacity = uint128(offer.availableCapacity.sub(size));
+
         emit NewAgreement(
+            agreementReference,
             dataReference,
             msg.sender,
             provider,
             size,
             billingPeriod,
+            billingPrice,
             msg.value
         );
     }
 
     /**
-    @notice extend the duration of the agreement.
-    @dev any safeMath operations in this function don't cause a deadlock, as a possible revert just means we can't prolong the agreement for the desired duration\
-    @param dataReference the reference to the to-be-pinned file
+    >> FOR CONSUMER
+    @notice deposits new funds to the Agreement.
+    @dev
+        - depositing funds to Agreement that is linked to terminated Offer is not possible
+    @param dataReference
     @param provider the address of the provider of the Offer.
     */
     function depositFunds(bytes32[] memory dataReference, address provider) public payable {
         bytes32 agreementReference = getAgreementReference(dataReference);
         Offer storage offer = offerRegistry[provider];
+        require(offer.availableCapacity != 0, "StorageManager: Offer for this Provider doesn't exist");
+
         Agreement storage agreement = offer.agreementRegistry[agreementReference];
-        require(offer.availableCapacity != 0, "StorageManager: Offer is terminated");
-        require(agreement.startDate != 0, "StorageManager: Agreement not active");
+        require(agreement.size != 0, "StorageManager: Agreement for this Offer doesn't exist");
+        require(agreement.lastPayoutDate != 0, "StorageManager: Agreement not active");
+
+        // TODO: [Q] Should we disallow depositing funds to Agreement whose chosen price does not exists anymore? What other options user has then?
         require(offer.billingPlans[agreement.billingPeriod] != 0, "StorageManager: price not available anymore");
```

Comment:
[Q] Should we disallow depositing funds to Agreement whose chosen price does not exist anymore? What other options user has then?

Reply:
Okey, but then when Provider removes billing plan he practically "terminates" all the Agreements. It is not "hard" termination (eq. after the current period) but "soft" (eq. will Agreement will run until it has funds, but can't be prolonged).

The consumer then has only two options. Either waits until the agreement runs out of funds or if he wants to move to a different billing plan ASAP, he can withdraw all available funds that are not reserved. Then he has to wait until Provider terminates the agreement with `payoutFunds` and only after then he can call again `newAgreement` with the new price.

This is not really optimal IMO. But agree that we should not allow it.

Reply:
We can notify consumers in the UI in case their current plans do not exist anymore. But I agree at the SC level we should not allow depositing more funds.